File: solanity_firebase_flutter/lib/models/comment.model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class CommentModel {
  final String id;
  final String userId;
  final String postId;
  final String userName;
  final String useravatarurl;
  final String text;
  final String? parentCommentId;
  final String? rootCommentId;
  final String? replyingToUser;
  final int repliesCount;
  final int likesCount;
  final Map<String, int> reactionsCount;
  final DateTime? createdAt;

  bool get isReply => parentCommentId != null;
  //get keyword  makes a variable getter, the value of that variable assigned according to logic written in function

  CommentModel({
    required this.id,
    required this.postId,
    required this.userId,
    required this.userName,
    required this.useravatarurl,
    required this.text,
    this.parentCommentId,
    this.rootCommentId,
    this.replyingToUser,
    this.repliesCount = 0,
    this.likesCount = 0,
    this.reactionsCount = const {},
    this.createdAt,
  });

  factory CommentModel.fromDoc(DocumentSnapshot<Map<String, dynamic>> doc) {
    final data = doc.data()!;
    return CommentModel(
      id: doc.id,
      postId: data['postId'],
      userId: data['userId'],
      userName: data['userName'],
      useravatarurl: data['useravatarurl'],
      text: data['text'],
      parentCommentId: data['parentCommentId'],
      rootCommentId: data['rootCommentId'],
      replyingToUser: data['replyingToUser'],
      repliesCount: data['repliesCount'] ?? 0,
      likesCount: data['likesCount'] ?? 0,
      reactionsCount: Map<String, int>.from(data['reactionsCount'] ?? {}),
      createdAt: (data['createdAt'] as Timestamp?)?.toDate(),
    );
  }
}

File: solanity_mern/server/src/models/comment.model.ts
import { Schema, model, Document, Types } from 'mongoose';

export interface IComment extends Document {
  _id: Types.ObjectId;
  postId: Types.ObjectId;
  userId: Types.ObjectId;
  userName: string;
  useravatarurl?: string | null;
  text: string;
  parentCommentId?: Types.ObjectId | null;
  rootCommentId?: Types.ObjectId | null;
  replyingToUser?: string | null;
  repliesCount: number;
  likesCount: number;
  reactionsCount: Map<string, number>;
  createdAt: Date;
}

const commentSchema = new Schema<IComment>(
  {
    postId: { type: Schema.Types.ObjectId, ref: 'Post', required: true, index: true },
    userId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    userName: { type: String, required: true },
    useravatarurl: { type: String, default: null },
    text: { type: String, required: true },
    parentCommentId: { type: Schema.Types.ObjectId, ref: 'Comment', default: null, index: true },
    rootCommentId: { type: Schema.Types.ObjectId, ref: 'Comment', default: null },
    replyingToUser: { type: String, default: null },
    repliesCount: { type: Number, default: 0 },
    likesCount: { type: Number, default: 0 },
    reactionsCount: { type: Map, of: Number, default: {} },
  },
  { timestamps: { createdAt: true, updatedAt: false } }
);

commentSchema.virtual('isReply').get(function (this: IComment) {
  return this.parentCommentId != null;
});

export const Comment = model<IComment>('Comment', commentSchema);
